packages/server/modules/core/services/ratelimiter.ts

```diff
@@ -0,0 +1,344 @@
+import express from 'express'
+import Redis from 'ioredis'
+import {
+  getRedisUrl,
+  getIntFromEnv,
+  isTestEnv
+} from '@/modules/shared/helpers/envHelper'
+import {
+  BurstyRateLimiter,
+  RateLimiterAbstract,
+  RateLimiterMemory,
+  RateLimiterRedis,
+  RateLimiterRes
+} from 'rate-limiter-flexible'
+import { TIME } from '@speckle/shared'
+import { AuthContext } from '@/modules/shared/authz'
+import { BaseError } from '@/modules/shared/errors'
+import { getIpFromRequest } from '@/modules/shared/utils/ip'
+import { Options } from 'verror'
+
+export class RateLimitError extends BaseError {
+  static defaultMessage =
+    'You have sent too many requests. You are being rate limited. Please try again later.'
+  static code = 'RATE_LIMIT_ERROR'
+
+  rateLimitBreached: RateLimitBreached
+
+  constructor(
+    rateLimitBreached: RateLimitBreached,
+    message?: string | null | undefined,
+    options: Options | Error | undefined = undefined
+  ) {
+    super(message ?? RateLimitError.defaultMessage, options)
```

Comment:
you rarely need to use `??` in JS, in this case it would be better to use `||` instead. `??` is only going to check for null while `||` is going to check for all falsy values